### PR TITLE
Improved irrep matrix code

### PR DESCRIPTION
**Pull request by attila-i-szabo**
This PR improves `FiniteGroup::_irrep_matrices`. The homegrown logic is replaced by [Dixon's irrep algorithm](https://www.ams.org/journals/mcom/1970-24-111/S0025-5718-1970-0280611-6/), which has two advantages:
* It still diagonalises a matrix of the same size as the group. However, the degenerate eigenvectors of this matrix span irreps directly, which saves us a lot of postprocessing. Accordingly, it runs [much faster](https://gist.github.com/attila-i-szabo/5ded177559efa53d2df5f5244c10d9b5).
* It ensures that real matrices are returned wherever possible. This involves redoing some calculations with real and complex matrices, which is a bit wasteful. However, other approaches I tried didn't perform any better, and this is quite transparent. It doesn't seem to be a major bottleneck, and it'd be more than offset by memory and time savings in the GCNN code @chrisrothUT and I are developing, which would be (as far as I can tell) the main point of use for this code.

### CODE REVIEW

**Comment by attila-i-szabo** [comment]
I think I can speed up the current bottleneck further, bear with me.

**Comment by PhilipVinc** [comment]
Flakehell complains about the following:
```python
➜ flakehell lint netket         

netket/utils/group/_group.py
    20:   1 F401 'functools.reduce' imported but unused [pyflakes]
  from functools import reduce
  ^
    26:   1 F401 'netket.utils.types.PyTree' imported but unused [pyflakes]
  from netket.utils.types import Array, PyTree
  ^
```

please fix them.

**Comment by PhilipVinc** [comment]
>It ensures that real matrices are returned wherever possible. This involves redoing some calculations with real and complex matrices, which is a bit wasteful. However, other approaches I tried didn't perform any better, and this is quite transparent. It doesn't seem to be a major bottleneck, and it'd be more than offset by memory and time savings in the GCNN code @chrisrothUT and I are developing, which would be (as far as I can tell) the main point of use for this code.

From my quick reading of the code, it seems to me that you return a list where you might be mixing real and complex-valued elements. is that right?
Are we sure that's a good idea? Especially if we plan to use this output in some models..

**Comment by PhilipVinc** [comment]
> Surprisingly, diagonalising big matrices isn't the bottleneck, it seems eigh is just optimised beyond our abilities!

I suspect that the bottleneck is the python looping.
If you want to really identify bottlenecks you should run the code several times under py-spy.
It will highlight the lines that take most time.

**Comment by attila-i-szabo** [comment]
> From my quick reading of the code, it seems to me that you return a list where you might be mixing real and complex-valued elements. is that right?
> Are we sure that's a good idea? Especially if we plan to use this output in some models..

It is a good idea for a couple of reasons:
* A lot of practically relevant space groups only have real irreps, so we should definitely use those as real
* The different irreps are logically separate and are handled separately in the model we're developing. Every irrep that can be handled using real rather than complex arithmetic should save some runtime. (It might turn out that vectorising over different irreps is better, but we can always promote the real matrices to complex.)
* Complex irreps come in conjugate pairs, so we're thinking if splitting them to real and imaginary parts would be better. (We might never do this though.)

> I suspect that the bottleneck is the python looping.

Most of the runtime is spent in that loop for sure. I don't have a lot of time for analysing it in detail, but I think we're good enough now that this routine consistently takes less time than `product_table`...

File: netket/utils/group/_group.py [hunk 1]
```diff
@@ -263,50 +263,80 @@ def character_table_readable(self) -> Tuple[List[str], Array]:
...
+            v = v[:, idx[first] : idx[first + 1]]
+            # v is now the basis of a single irrep: project the regular rep onto it
             irreps.append(
                 prune_zeros(
-                    np.einsum("gi,ghj ->hij", w.conj(), w[true_product_table, :])
+                    np.einsum("gi,ghj ->hij", v.conj(), v[true_product_table, :])
                 )
             )
 
         return irreps
```
Comment (chrisrothUT) [inline netket/utils/group/_group.py:337]: Can you make sure the dtype returned is a float if the irreps are real and otherwise complex? Just to make sure we don't end up in situations where complex numbers are used when they don't need to be

**Review by chrisrothUT — APPROVED** [review]

File: netket/utils/group/_group.py (hunk 1 above)
Comment (chrisrothUT) [reply netket/utils/group/_group.py:337]: Nevermind, I looked more closely and this should be the case, carry on

File: netket/utils/group/_group.py [hunk 2]
```diff
@@ -263,50 +263,80 @@ def character_table_readable(self) -> Tuple[List[str], Array]:
...
+            s = np.random.standard_normal(len(self))
+            s = s[self.product_table[np.ix_(self.inverse, self.inverse)]]
+            proj = self.character_table().conj() @ s @ vs
+            starting_idx = list(starting_idx) + [len(self)]
+            return v, starting_idx, proj
+
+        eigen = {}
+        if np.any(frob == 1):
+            # real irreps: start from a real symmetric invariant matrix
+            e = np.random.standard_normal(len(self))
```
Comment (chrisrothUT) [inline netket/utils/group/_group.py:313]: Is there any particular reason we need randomness in this algorithm? Is the idea to avoid an accidental symmetry in the irreps? And if so, we can still make sure the algorithm gives the same irres every time

Comment (attila-i-szabo) [reply netket/utils/group/_group.py:313]: > Is the idea to avoid an accidental symmetry in the irreps?

Exactly. The eigenspaces of the matrix `E` are all reps, but we want to make sure they're irreps too.

> My fear is that you save a model that uses a particular set of irreps and then you load it with a different set of irreps and it doesn't work

This is basically the reason we chose to keep defining the GCNN kernels in terms of group elements rather than the group FT. We can't guarantee we return the same irrep matrices every time, as they are equivalent up to unitary basis rotations, and the eigenvectors `eigh` returns in degenerate eigenspaces are essentially random.

> This doesn't matter for convolution but for projection it's important

Projections only use the characters, which are invariant under these unitary rotations.

**Comment by attila-i-szabo** [comment]
One more thing @PhilipVinc. We've observed when debugging the tests of #724 that `character_table` produces much less accurate characters than the traces of these matrices, so it would make sense to recompute characters at the end of this routine if it gets called. But is there a way to replace cached values in a `dataclass`?

File: netket/utils/group/_group.py (hunk 2 above)
Comment (chrisrothUT) [reply netket/utils/group/_group.py:313]: > Projections only use the characters, which are invariant under these unitary rotations.

We may want to implement a learned projection into the degenerate subspace for irreps with d > 1. I've been finding the character projections don't work particularly well for higher dimensional representations

**Comment by femtobit** [comment]
> But is there a way to replace cached values in a dataclass?

If the value can change, it is not a proper (frozen) dataclass, that would be a bug in my view. Also, too much magic happening under the hood. If `character_table` can be made more accurate using this routine, but calling it is optional, I'd just add an option to `character_table`.

File: netket/utils/group/_group.py (hunk 2 above)
Comment (femtobit) [reply netket/utils/group/_group.py:313]: > We can't guarantee we return the same irrep matrices every time, 

At least this function should take a `key` then, so that when used with the same seed the results also match even though pseudorandom numbers are used.

We got most other parts of NetKet to give reproducible results when used with the same seed, this should probably not be an exception.

Comment (attila-i-szabo) [reply netket/utils/group/_group.py:313]: > > Projections only use the characters, which are invariant under these unitary rotations.
> 
> We may want to implement a learned projection into the degenerate subspace for irreps with d > 1. I've been finding the character projections don't work particularly well for higher dimensional representations

What do you mean? The character projection is *the* valid projection into an irrep subspace...

> > We can't guarantee we return the same irrep matrices every time,
> 
> At least this function should take a `key` then, so that when used with the same seed the results also match even though pseudorandom numbers are used.
> 
> We got most other parts of NetKet to give reproducible results when used with the same seed, this should probably not be an exception.

Fair enough, although that means we switch to `jax.random` here, right?

Comment (femtobit) [reply netket/utils/group/_group.py:313]: Hm, that would probably be easiest.

You could also make a random seed from the key and use that to initialize a numpy RNG, but that seems to just add more steps to get the same result.

@PhilipVinc Do you agree?

Comment (attila-i-szabo) [reply netket/utils/group/_group.py]: Okay, this has also been a bit of an issue for `character_table`, so I switched that to `jax.random` too. Also, some of the logic is now moved to JAX, which might help speed up the linear algebra (and we need the output in JAX code anyway). However, it's now a bit of a mess between JAX and old numpy, and moving it entirely to JAX would be inconvenient, so I wonder if sticking to the latter (apart from random numbers) might be better?

Comment (attila-i-szabo) [reply netket/utils/group/_group.py]: Cool. For sake of simplicity I only switched the random numbers to JAX, but we might want to put some computations on the GPU too?

Comment (femtobit) [reply netket/utils/group/_group.py]: In principle I am happy just drawing the random numbers using JAX for consistency with the rest of the code.

Of course I wouldn't mind a potential performance gain from using JAX beyond that. If you want to try and benchmark this, why not. OTOH, it's not crucial for this PR as far as I'm concerned.

Comment (attila-i-szabo) [reply netket/utils/group/_group.py]: The bottleneck seems to be doing relatively small linear algebra calculations in a big for loop, and I can't think of any remotely memory-friendly way of vectorising it. I.e. I doubt GPUs would bring a lot of improvement, so I rather won't spend much dev time on sorting out a clean JAX implementation...

File: netket/utils/group/_group.py [hunk 3]
```diff
@@ -263,50 +264,79 @@ def character_table_readable(self) -> Tuple[List[str], Array]:
...
-        # The regular representation contains d copies of each d-dimensional irrep
-        # all of which return the same eigenvalues with eigenvectors that
-        # correspond to one another in the bases of the several irrep copies
-        e, vs = np.linalg.eig(random(len(self))[self.product_table])
-        e = np.stack((e.real, e.imag))
-        # we only need one eigenvector per eigenvalue
-        _, idx = np.unique(comparable(e), return_index=True, axis=1)
-        vs = vs[:, idx]
-        # We will nedd the true product table as well
+    def _irrep_matrices(self, key=jax.random.PRNGKey(0)) -> List[Array]:
```
Comment (PhilipVinc) [inline netket/utils/group/_group.py:267]: I need to think how to properly do this.
But I fear this default prngkey here is a sure recipe for hidden bugs...

You are basically saying that unless passed a key, this will always use the same random seed.
Which might make sense, maybe, but I'm unsure...

File: netket/utils/group/_group.py [hunk 4]
```diff
@@ -263,50 +264,79 @@ def character_table_readable(self) -> Tuple[List[str], Array]:
...
+        squares = np.diag(true_product_table)
+        frob = np.array(
+            np.rint(
+                np.sum(self.character_table()[:, squares], axis=1).real / len(self)
+            ),
+            dtype=int,
+        )
+
+        def random(n, key, dtype=float):
+            return np.asarray(jax.random.normal(key, (n,), dtype))
```
Comment (PhilipVinc) [inline netket/utils/group/_group.py:287]: I don't understand why you want to use jax if then you reconvert to numpy

File: netket/utils/group/_group.py (hunk 3 above)
Comment (PhilipVinc) [reply netket/utils/group/_group.py:267]: Moreover, this is a property. So you cannot pass the optional kwarg, and it will _always_ have the default value.

Can you try to explain me why exactly you need random numbers here, and what are you trying to achieve?

Comment (PhilipVinc) [reply netket/utils/group/_group.py:267]: can't we fix the random vector? 
Is there a reason why that _random_ vector must be completely random?

Comment (attila-i-szabo) [reply netket/utils/group/_group.py:267]: The algorithm works by splitting the regular representation (which is known to contain every irrep) into constituent irreps and picking out one for every irrep listed in the character table. This it does by constructing a Hermitian matrix (`e` after the second line of `invariant_subspaces`) that commutes with every matrix of the regular rep. This means that the subspaces spanned by degenerate eigenvectors transform as some rep of the group. We need to make sure that these are all irreps, i.e., there are no accidental degeneracies. The simplest way of doing that is using a random `e` matrix (as long as it commutes with the regular rep matrices). 

The JAX random number generator was included at the request of @femtobit (IDK where that conversation went) to make sure you always get the same irrep matrices, rather than something related by a unitary transformation. IMO we should fix that key once and for all, rather than trying to pass it. (We could also cache a bunch of random numbers from random.org, but this is more transparent.)

File: netket/utils/group/_group.py (hunk 4 above)
Comment (attila-i-szabo) [reply netket/utils/group/_group.py:287]: because the JAX PRNG is more easily reproducible

File: netket/utils/group/_group.py (hunk 3 above)
Comment (attila-i-szabo) [reply netket/utils/group/_group.py:267]: BTW there's a similar pattern in `character_table_by_class` but there it matters less because the output is deterministic. For irrep matrices, you have the freedom of conjugating all matrices in an irrep with some unitary matrix; randomising the matrix randomises this unitary transformation.

Comment (femtobit) [reply netket/utils/group/_group.py:267]: > They need to look random to make sure we don't have accidental degeneracies.

But in principle, it should be sufficient to pick a specific random vector, right? I don't mean putting it explicitly in the code, just not exposing an option to change the key and hard-code a specific one (or use numpy with a specific seed, but the JAX API is a bit more comfortable here).

Comment (attila-i-szabo) [reply netket/utils/group/_group.py:267]: Exactly, I'm not sure why I added the key argument, it only makes things less reproducible (and doesn't play well with caching as @PhilipVinc pointed out)

File: netket/utils/group/_group.py [hunk 5]
```diff
@@ -204,7 +204,9 @@ def character_table_by_class(self) -> Array:
         # since we only want a random linear combination, we forget about the
         # constant |S| and only divide each column through with the appropriate |T|
         class_matrix = (
-            classes @ np.random.uniform(size=len(self))[self.product_table] @ classes.T
+            classes
+            @ random(len(self), jax.random.PRNGKey(0))[self.product_table]
```
Comment (femtobit) [inline netket/utils/group/_group.py:208]: Yes, that's what is was thinking.

I suggest adding a comment here so people looking at this in the future understand what the purpose of the random vector is.

Comment (attila-i-szabo) [reply netket/utils/group/_group.py:208]: I added some more comments outlining the whole algorithm at the start (there was already some here!) 
Could you draft something if you think more is needed?

File: netket/utils/group/_group.py (hunk 4 above)
Comment (PhilipVinc) [reply netket/utils/group/_group.py]: to make this code less confusing, if this is a numpy code, I would use numpy new random number generator interface which is essentially the same as jax PRNG system.
In the not-so-distant future, Numba will also support those and so we could jit everything.

```python
rng = numpy.random.default_rng(SEED)
out = rng.normal((n,), dtype=dtype)
```

This is perfectly reproducible.

cc @femtobit, what do you think?

File: netket/utils/group/_group.py (hunk 5 above)
Comment (PhilipVinc) [reply netket/utils/group/_group.py:208]: See my comment above.
To make it more 'transparent' I'm ok with fixing the PRNG seed but then let's just use numpy.*
It will also be faster.

File: netket/utils/group/_group.py (hunk 4 above)
Comment (femtobit) [reply netket/utils/group/_group.py]: I'm happy with the current state, but I don't mind either way. Note that in case we want to numba-jit this later we could also just change it then, it's trivial to do after all.

Comment (PhilipVinc) [reply netket/utils/group/_group.py]: It's not entirely necessary, but I'd prefer not to have jax if it's not needed.
I fear that we won't remember why we're using it and it's just confusing.

Comment (gcarleo) [reply netket/utils/group/_group.py]: I actually agree, I think that this part on the symmetries is somehow self-contained and should not depend too much on the autodiff things/jax

**Review by PhilipVinc — CHANGES_REQUESTED** [review]
it's good for me if you remove jax and use numpy rng devices.

**Comment by attila-i-szabo** [comment]
Done – good to know numpy developed this kind of RNG!

BTW, I'm not sure jitting buys us much on these routines – their outputs are cached, so they are not to be run more than once anyway

**Review by PhilipVinc — APPROVED** [review]
